### DISCUSSION TITLE
Netlify redirects file

### DISCUSSION THREAD

**Discussion by corbob** [statiqdev/Discussions#99] [Q&A] [answered]
Hello,

I'm trying to get the `_redirects` file from my `input` directory into Netlify. Following the directions here: https://www.statiq.dev/web/content-and-data/content/redirects I've created an `appsettings.json` file that looks like this: 

```json
{
    "NetlifyRedirects": "true"
}
```

but my deployment to Netlify doesn't appear to include the `_redirects` file. I have attempted to deploy through `dotnet run -- deploy` with the proper settings in `appsettings.json` to a successful deployment still lacking the `_redirects` file. Is there a step I'm missing somewhere?

**Comment by gep13** [comment, marked as answer]
What that part of the docs is telling you is that Statiq will look for properties on specific documents, and it will create redirects (either in the form of pages with metadata redirects, or by creating a `_redirects` file for you automatically, when the `NetlifyRedirects` is set.

You can see these examples of what I am referring to in the Chocolatey docs repository:

https://github.com/chocolatey/docs/search?q=RedirectFrom

Each one of these attributes results in a redirect page being created in the output folder when we build the site.

As an example, go into the `about.md` file in your repository, and add the following:

```
RedirectFrom: gary
```

Run the build, and a new `_redirects` file will be created in the output folder with the following content:

<img width="410" alt="2021-10-14_19-42-34" src="https://user-images.githubusercontent.com/1271146/137377062-da0d8064-1d4b-4fd0-b020-1de50ac87f0e.png">

So, in order to get the content that you want in the _redirects file, we need to add a RedirectFrom attribute in the document that the root of the rss is currently being created from.  However, this is an index.yaml file:

https://github.com/corbob/coryknox.dev/blob/master/input/blog/rss/index.yaml

And when I tried adding the RedirectFrom attribute to here, it didn't work.

We would need @daveaglick to chime in here with what needs to be done here to make this work.

**Reply by daveaglick** [reply]
This is exactly right - with `NetlifyRedirects` set, Statiq will attempt to create it's own `_redirects` file from the front matter of documents. So adding that front matter rather than creating your own redirects file is one way to handle it.

On the other hand, you may want the additional control of writing your own redirects file. The reason yours isn't being copied to output is because it starts with an underscore, and underscore files are ignored by default (from [the docs](https://www.statiq.dev/web/content-and-data/)):
> Files should be placed in the input folder (or a sub-folder) and the InputFiles setting controls what files from the input folder are loaded. It's set to **/{!_,}* by default which loads all files in any input directory unless it starts with an underscore _.

So the trick is to write the redirects file in a way that it does end up in the output folder with the name `_redirects`. As with everything in Statiq, there's a few ways to accomplish this (and make sure you _don't_ set `NetlifyRedirects` if you're going one of these routes with your own redirects file).

The first way I can think of is to not use an underscore initially and specify a destination path in the front matter of a redirects file named `redirects.html` (without the underscore, but with an `.html` extension - more on that below):
```
DestinationPath: _redirects
ContentType: Asset
MediaType: text/plain
---
# Redirects from what the browser requests to what we serve
/home              /
/blog/my-post.php  /blog/my-post
/news              /blog
```
Note the `.html` extension in the file name - it'll get removed by the `DestinationPath` front matter when the file is output, but at least for now we need it to trigger front matter processing since we only look at front matter for certain types of files. [There's a new issue I opened](https://github.com/statiqdev/Statiq.Web/issues/967) for a more generic `.statiq` extension that should do the trick in the future (this is now the third time I've needed this, so it'll probably get implemented soon).

Another option would be to add an `InputFiles` path that includes your `_redirects` file directly. Add this to your `settings.yml` file (note that we also need to add the default pattern or else you'll _only_ be picking up the redirects file):
```
InputFiles:
  - "**/{!_,}*"
  - _redirects
```

Do both those options make sense (and particularly the reason why the `_redirects` file wasn't picked up in the first place)? Now that it's hopefully clear what's going on and why, I'd also be interested in any suggestions on how to improve usability for something like this.
